AdaptivePoolingAllocator EventLoop Magazine's affinity

franz1981 opened the pull request:
Motivation:

AdaptivePoolingAllocator uses some shared Magazines hoping thread's id distribution to improve contended behaviour, but event loops threads can just uses their own Magazine(s) and save 2 atomic operations (write lock/unlock) in the hot path, further improving performance.

Modification:

Implements dedicated event loop Magazines

Result:

Better allocation performance for event loop threads

franz1981 commented:
@chrisvest @normanmaurer Similarly to what I've done for https://github.com/quarkusio/quarkus/blob/ca27e0c29d9ae048227da8118c61ef3215fe6266/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/AppendBuffer.java#L21 we can use a JDK 8 `MethodHandle::invokeExact` and explicitly save ThreadLocal(s) for virtual thread(s) too, but I didn't pushed this here.

chrisvest commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -259,9 +313,13 @@ private static class AllocationStatistics extends StampedLock {
-        private AllocationStatistics(AdaptivePoolingAllocator parent) {
+        private AllocationStatistics(AdaptivePoolingAllocator parent, boolean fastThreadLocalStatistics) {
             this.parent = parent;
+            this.lock = !fastThreadLocalStatistics? new StampedLock() : null;
```
I'd prefer to keep extending StampedLock and predicate on a boolean instead.

chrisvest commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -96,25 +100,56 @@ final class AdaptivePoolingAllocator {
+            for (int i = 0; i < mags.length; i++) {
+                mags[i] = new Magazine(this);
+            }
+            magazines = mags;
```
Unused magazines take up very little memory. Let's just always initialize them like before.
The `useFastThreadLocalMagazines` should only enable a short-cut path in `allocate()` and determine the allocation of the thread-local.

chrisvest commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -159,25 +194,38 @@ ByteBuf allocate(int size, int maxCapacity) {
+            Magazine mag = fastThreadLocalMagazines.get();
+            return mag.allocate(size, sizeBucket, maxCapacity, buf);
+        }
+        // Compute outside of Magazine lock for better ILP.
```
This comment got divorced from `int sizeBucket = AllocationStatistics.sizeBucket(size);` on line 197.

chrisvest commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -159,25 +194,38 @@ ByteBuf allocate(int size, int maxCapacity) {
-                        mag.unlockWrite(writeLock);
+            // mags.length == 0 => mask == -1:
+            // no need to increment expansions here if we're empty: we really want to allocate a magazine!
+            if (mask >= 0) {
```
Let's keep this as before and not use `EMPTY_MAGAZINES`.

chrisvest commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -300,8 +358,15 @@ private void rotateHistograms() {
+            } else {
+                tmpPrefChunkSize = prefChunkSize;
+                // try avoid creating useless garbage here: this is always supposed to be called from the same thread!
+                parent.liveFastThreadLocalMagazines.forEach(this);
```
I think thread-local magazines should probably not share their histograms at all, and only compute a local preferred chunk size. They are only ever allocating for one thread, so I think it'd be fine to optimize for whatever that thread prefers.

chrisvest reviewed (commented):
Had some initial comments.

franz1981 commented:
Thanks @chrisvest and sorry for arriving late

franz1981 commented:
I'm sending an additional commit to further improve performance @chrisvest, but i didn't yet added specific tests or benchmarks on this, yet.

franz1981 commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -160,6 +179,14 @@ ByteBuf allocate(int size, int maxCapacity) {
         int sizeBucket = AllocationStatistics.sizeBucket(size); // Compute outside of Magazine lock for better ILP.
+        FastThreadLocal<Magazine> fastThreadLocalMagazines = this.fastThreadLocalMagazines;
+        if (fastThreadLocalMagazines != null && currentThread instanceof FastThreadLocalThread) {
+            // TODO this is opinionated: ignoring on purpose non FastThreadLocalThread(s):
```
@chrisvest I don't have yet strong opinion on this...

I see 3 options:
1. uses a `MethodHandle::invokeExact` to check for virtual threads current thread: this will help to use `FastThreadLocal`s with no fear as we do for https://github.com/netty/netty/blob/f9e20ac0490d2877c81def47a7d4046e06d9f410/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L516-L524 helping users which relies on something like `io.netty.allocator.useCacheForAllThreads`, to correctly distinguish between event loop threads NOT using `FastThreadLocalThread`, event loop threads extending `FastThreadLocalThread` and non event loop threads still extending `FastThreadLocalThread`
2. leave it as it is, but it means not helping users which are abusing `FastThreadLocalThread`s for non-event lop threads 
3. like 2, but adding a subsequent check (configurable), to disable non-shareable `Magazine`s for `FastThreadLocalThread`s which are not event loop ones; this won't help users which have event loop threads with custom threads

I'm sending a commit for this last option, but feel free to suggest something different.

franz1981 commented:
Perf wise, I've profiled `ByteBufAllocatorBenchmark` and given that is an uncontended case, I see that the dominant costs are instead the many others atomic ops to increase/decrease the ref cnt shared on the chunk buffer; due to this the uncontended case with 256 bytes and direct buffer have just a small improvement i.e. 23 M ops/sec ->  25 M ops/sec

Where instead I think it would very beneficial, is with the concurrent scenario in the real world, because there won't be any retry due to contention or expansion going on for event loop magazines, given that are not shreable, but I don't have any performance test around such yet.

franz1981 commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -158,8 +190,14 @@ ByteBuf allocate(int size, int maxCapacity) {
         int sizeBucket = AllocationStatistics.sizeBucket(size); // Compute outside of Magazine lock for better ILP.
+        FastThreadLocal<Magazine> fastThreadLocalMagazines = this.fastThreadLocalMagazines;
+        if (fastThreadLocalMagazines != null && currentThread instanceof FastThreadLocalThread &&
+            (DEFAULT_USE_CACHED_MAGAZINES_FOR_NON_EVENT_LOOP_THREADS || ThreadExecutorMap.currentExecutor() != null)) {
```
I'm not very convinced about having `ThreadExecutorMap.currentExecutor()` in the hot path, really...

franz1981 replied:
Sent a203ba3 to address this, as usual, in a very ugly way.

franz1981 commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -97,20 +110,55 @@ final class AdaptivePoolingAllocator {
+            cachedMagazines = new FastThreadLocal<Object>() {
+                @Override
+                protected Object initialValue() {
+                    if (cachedMagazinesNonEventLoopThreads || ThreadExecutorMap.currentExecutor() != null) {
```
I have to find a way to benchmark this `ThreadExecutorMap::currentExecutor` call in the hot path @chrisvest 

I don't like it, but I want to help users which uses `FastThreadLocalThread` for non-event loop threads

normanmaurer commented:
@franz1981 please fix

normanmaurer commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -97,20 +110,55 @@ final class AdaptivePoolingAllocator {
     private final StampedLock magazineExpandLock;
     private volatile Magazine[] magazines;
+    private final FastThreadLocal<Object> cachedMagazines;
+    private final CopyOnWriteArraySet<Magazine> liveCachedMagazines;
```
```suggestion
    private final Set<Magazine> liveCachedMagazines;
```

normanmaurer commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -97,20 +110,55 @@ final class AdaptivePoolingAllocator {
+                   magazineCaching == MagazineCaching.FastThreadLocalThreads;
+            final boolean cachedMagazinesNonEventLoopThreads =
+                    magazineCaching == MagazineCaching.FastThreadLocalThreads;
+            CopyOnWriteArraySet<Magazine> liveMagazines = new CopyOnWriteArraySet<>();
```
```suggestion
            Set<Magazine> liveMagazines = new CopyOnWriteArraySet<>();
```

franz1981 commented:
@normanmaurer hope to have addressed the comments

I still believe, for the hot path vs event loops we could do **much** better, but I need to better understand how this work: the `ChunkByteBuf` uses a costy `retain` method to bump the allocation out of it, but in theory this should be fully uncontended, because it is either protected by a lock or not shared: maybe there is something we can do there to split the ref cnt for `ChunkByteBuf` into 2 parts, making the `retain` much less costy, using something which doesn't imply strong memory barriers, but I need to better understand the allocator, I think

chrisvest commented:
The `ChunkByteBuf.retain()` call coordinates with the deallocations of allocated `ByteBuf` instances, which can happen in any thread because buffers are not confined.

franz1981 commented:
Yep @chrisvest that's why I proposed 2 separate counters (up and down): so they can be updated with different ordering mode. We usually decide for the subsequent deallocation of the chunk when its capacity is exhausting (on the last adaptive buffer we slice out of it - which is still in the allocation path), and maybe this can be used at our advantage. I have to better think if I am saying something just not possible, but I will give it a shot in the next days or in another pr.

This one, as it is already deliver a decent improvement, which can likely get better under magazine sharing (which cannot happen anymore  between event loops).

chrisvest commented on buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:
```diff
@@ -97,20 +110,55 @@ final class AdaptivePoolingAllocator {
     private final Queue<ChunkByteBuf> centralQueue;
     private final StampedLock magazineExpandLock;
     private volatile Magazine[] magazines;
+    private final FastThreadLocal<Object> cachedMagazines;
```
Maybe `threadLocalMagazine` is a better name?

chrisvest reviewed (approved):
One suggestion, but otherwise looks good. Let's look at the `ChunkByteBuf` reference count in a separate PR.

franz1981 commented:
@chrisvest this is what I meant (@normanmaurer too):

https://github.com/franz1981/netty/commit/724d7540d3d2b955e47b48ddad8dbe0ba9c6a9db


as you can see it is relaxing the ref cnt behaviour, but it shouldn't be more unsafe than what we have right now, given the constrains of the algorithm and it would enable super fast-path event loops allocations.

wdyt?

If it looks good I can port it on this branch as well...

chrisvest commented:
@franz1981 I think that looks interesting. I'd like to think it through independently and see what the build says, so I still prefer it being a separate PR.

franz1981 commented:
ready to go @normanmaurer too?

I'll send the ref cnt changes in a separate PR, which speedup quite a lot the event loops allocation ;)

franz1981 commented:
@chrisvest I can send some more test or a better benchmark for this or it is ready to go?

chrisvest commented:
I think this is ready. Was just waiting to see if Norman had anything, but I think we can merge it

normanmaurer commented:
Cherry-picked into 4.2 ... @franz1981 do you mind opening a PR against main ?

franz1981 commented:
Will take tomorrow folks, dealing with some other things these days..and will send the ref cnt proposal @chrisvest which clearly relax something but for a very real positive performance impact, but I can make it opt-in (I believe)
